esrally/track/params.py

```diff
@@ -901,38 +898,86 @@ def __next__(self):
         try:
             docs_in_batch = 0
             while docs_in_batch < self.batch_size:
-                docs_in_bulk, bulk = self.read_bulk()
+                try:
+                    docs_in_bulk, bulk = self.read_bulk()
+                except StopIteration:
+                    break
                 if docs_in_bulk == 0:
                     break
                 docs_in_batch += docs_in_bulk
-                batch.append((docs_in_bulk, bulk))
+                batch.append((docs_in_bulk, "".join(bulk)))
             if docs_in_batch == 0:
                 raise StopIteration()
             return self.index_name, self.type_name, batch
         except IOError:
             logging.getLogger(__name__).exception("Could not read [%s]", self.data_file)
 
-    def read_bulk(self):
-        docs_in_bulk = 0
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.file_source.close()
+        return False
+
+
+class MetadataIndexDataReader(IndexDataReader):
+    def __init__(self, data_file, batch_size, bulk_size, file_source, action_metadata, index_name, type_name):
+        super().__init__(data_file, batch_size, bulk_size, file_source, index_name, type_name)
+        self.action_metadata = action_metadata
+        self.action_metadata_line = None
+
+    def __enter__(self):
+        super().__enter__()
+        if self.action_metadata.is_constant:
+            _, self.action_metadata_line = next(self.action_metadata)
+            self.read_bulk = self._read_bulk_fast
+        else:
+            self.read_bulk = self._read_bulk_regular
+        return self
+
+    def _read_bulk_fast(self):
+        """
+        Special-case implementation for bulk data files where the action and meta-data line is always identical.
+        """
         current_bulk = []
-        for action_metadata_item, document in zip(self.action_metadata, self.file_source):
+        # hoist
```

Comment:
Firstly I had to look this up as my understanding this is mostly JS terminology whereas we in Python we talk about block scoping.

But what does this comment refer at example? Which variable, in which block, is getting hoisted?

Reply:
I know that term from JVM optimizations. In any case we ensure that the field access is turned into a local variable access because it is used in the loop on the hot code path and this is what I was referring to here.

Reply:
I see. For a person like me it doesn't add any additional info (seems obvious from the implementation) but if it's valuable to someone else, that's fine.

Reply:
I did not mean to confuse with the comment. Thanks for the feedback. Let's remove the comment then if it is not useful.